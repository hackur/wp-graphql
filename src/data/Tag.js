import { toGlobalId } from 'graphql-relay';
import Dataloader from 'dataloader';
import fetchData from 'data/utils';

// Dataloader expects IDs that can be read by the REST API

const path = process.env.WP_TAGS_ENDPOINT || 'wp/v2/tags';
const tagLoader = new Dataloader(ids =>
  fetchData(path, {
    qs: { include: ids, per_page: 100, orderby: 'include' },
  }).then(({ data: { body } }) => body)
);
const slugLoader = new Dataloader(slugs =>
  fetchData(path, { qs: { slug: slugs, per_page: 100 } })
    .then(({ data: { body } }) => body)
    // the REST API does not order by FIELD(slug, ....) yet
    .then(tags => slugs.map(slug => tags.find(tag => slug === tag.slug)))
);

class Tag {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return path;
  }

  static async load(id) {
    const data = await tagLoader.load(id);
    return data ? Object.assign(new Tag(), data) : null;
  }

  static async loadMany(ids) {
    const data = await tagLoader.loadMany(ids);
    return data ? data.map(entry => Object.assign(new Tag(), entry)) : null;
  }

  static async loadBySlug(slug) {
    const data = await slugLoader.load(slug);
    return data ? Object.assign(new Tag(), data) : null;
  }
}

export default Tag;
