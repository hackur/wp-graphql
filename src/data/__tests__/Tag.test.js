import Tag from 'data/Tag';

describe('Test Tag data access', () => {
  test('Get endpoint', () => {
    expect(Tag.getEndpoint()).toMatchSnapshot();
  });

  test('Load a tag', async () => {
    const tag = await Tag.load(13);
    expect(tag.getID()).toMatchSnapshot();
    expect(tag).toMatchSnapshot();
  });

  test('Load a tag by slug', async () => {
    const tag = await Tag.loadBySlug('dirty-projectors');
    expect(tag).toMatchSnapshot();
  });

  test('Load tags', async () => {
    const tag = await Tag.loadMany([13, 17]);
    expect(tag).toMatchSnapshot();
  });
});
