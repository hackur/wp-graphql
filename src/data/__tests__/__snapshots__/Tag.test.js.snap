// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Tag data access Get endpoint 1`] = `"wp/v2/tags"`;

exports[`Test Tag data access Load a tag 1`] = `"VGFnOjEz"`;

exports[`Test Tag data access Load a tag 2`] = `
Tag {
  "id": 13,
}
`;

exports[`Test Tag data access Load a tag by slug 1`] = `
Tag {
  "slug": "dirty-projectors",
}
`;

exports[`Test Tag data access Load tags 1`] = `
Array [
  Tag {
    "id": 13,
  },
  Tag {
    "id": 17,
  },
]
`;
